### ISSUE TITLE
[Support]: Use hwaccel with nvidia to record.

### ISSUE THREAD

**Issue by caublet** [blakeblackshear/frigate#6184]
### Describe the problem you are having

I'm able to use my nvidia GPU to view streams no problem. The issues comes from when a recording starts; encoding of recording is using CPU.. My GTX1050 is capable of encoding x264 and I would like to use it to lower cpu usage.

nvidia-smi show all 4 cams decoding using nvidia : 
```
Thu Apr 20 13:59:29 2023
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 515.48.07    Driver Version: 515.48.07    CUDA Version: 11.7     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  On   | 00000000:01:00.0 Off |                  N/A |
|  0%   48C    P0    N/A / 120W |    243MiB /  2048MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A   4098201      C   ffmpeg                             64MiB |
|    0   N/A  N/A   4098209      C   ffmpeg                             64MiB |
|    0   N/A  N/A   4098214      C   ffmpeg                             55MiB |
|    0   N/A  N/A   4098227      C   ffmpeg                             55MiB |
+-----------------------------------------------------------------------------+
```

### Version

0.12.0-DA3E197

### Frigate config file

```yaml
go2rtc:
  streams:
#####      
    Side:
      - rtsp://admin:{FRIGATE_RTSP_PASSWORD}@192.168.1.23:554/Streaming/Channels/1
      - "ffmpeg:Side#audio=aac"
    SideSub:
      - rtsp://admin:{FRIGATE_RTSP_PASSWORD}@192.168.1.23:554/Streaming/Channels/2
#####      
    Back:
      - rtsp://admin:{FRIGATE_RTSP_PASSWORD}@192.168.1.24:554/Streaming/Channels/1
      - "ffmpeg:Back#audio=aac"
    BackSub:
      - rtsp://admin:{FRIGATE_RTSP_PASSWORD}@192.168.1.24:554/Streaming/Channels/2
#####      
    Ssol:
      - ffmpeg:rtsp://frigate:{FRIGATE_RTSP_PASSWORD}@192.168.1.21:554/Streaming/Channels/1
      - "ffmpeg:Ssol#audio=aac"
    SsolSub:
      - ffmpeg:rtsp://frigate:{FRIGATE_RTSP_PASSWORD}@192.168.1.21:554/Streaming/Channels/2
#####      
    Ssolfenetre:
      - ffmpeg:rtsp://frigate:{FRIGATE_RTSP_PASSWORD}@192.168.1.25:554/Streaming/Channels/1
      - "ffmpeg:Ssolfenetre#audio=aac"
    SsolfenetreSub:
      - ffmpeg:rtsp://frigate:{FRIGATE_RTSP_PASSWORD}@192.168.1.25:554/Streaming/Channels/2
#####      
  webrtc:
    candidates:
      - 192.168.1.15:8555
      - stun:8555
##################### CAMERAS ###################
cameras:
  Side:
    ffmpeg:
      output_args:
        record: preset-record-generic-audio-copy
      inputs:
        - path: rtsp://127.0.0.1:8554/Side
          input_args: preset-rtsp-restream
          roles:
            - record
        - path: rtsp://127.0.0.1:8554/SideSub
          input_args: preset-rtsp-restream
          roles:
            - detect
    detect:
      fps: 6
      width: 640
      height: 360
    objects:
      track:
        - person
        - car
      filters:
        person:
          min_score: 0.35
          threshold: 0.5
    record:
      enabled: True
      retain:
        days: 7
        mode: motion
      events:
        retain:
          default: 60
          mode: active_objects
        required_zones:
          - parking
          - sidedoor        
    snapshots:
      enabled: true
      timestamp: false
      bounding_box: true
      retain:
        default: 60
      required_zones:
        - parking
        - sidedoor        
    zones:
      parking:
        coordinates: 85,221,131,215,119,254,155,254,165,360,42,360,42,248
        objects:
          - car
          - person
      sidedoor:
        coordinates: 159,247,228,224,308,233,316,288,289,297,296,360,170,360
        objects:
          - person
    motion:
      mask:
        - 308,233,323,262,330,360,640,360,640,0,0,0,0,360,38,360,40,236,88,210,137,210,129,250,224,212
#########################################################################
  Back:
    ffmpeg:
      output_args:
        record: preset-record-generic-audio-copy
      inputs:
        - path: rtsp://127.0.0.1:8554/Back
          input_args: preset-rtsp-restream
          roles:
            - record
        - path: rtsp://127.0.0.1:8554/BackSub
          input_args: preset-rtsp-restream
          roles:
            - detect
    detect:
      fps: 6
      width: 640
      height: 360
    objects:
      track:
        - person
        - cat
      filters:
        person:
          min_score: 0.35
          threshold: 0.5
        cat:
          min_score: 0.35
          threshold: 0.5
    record:
      enabled: True
      retain:
        days: 7
        mode: motion
      events:
        retain:
          default: 60
          mode: active_objects
        required_zones:
          - pool
          - deck
          - door
          - poolgate
          - shed
    snapshots:
      enabled: true
      timestamp: false
      bounding_box: true
      retain:
        default: 60
      required_zones:
          - pool
          - deck
          - door
          - poolgate
          - shed
    zones:
      pool:
        coordinates: 408,187,439,236,499,226,501,191,485,145,450,103,415,83,392,90,385,129
        objects:
          - person
      deck:
        coordinates: 74,257,313,295,304,129,205,101,113,80,105,108,145,125,133,190,90,181
        objects:
          - person
      backdoor:
        coordinates: 86,177,126,187,141,125,102,111
        objects:
          - person
      poolgate:
        coordinates: 423,283,540,271,528,227,411,246
        objects:
          - person
      shed:
        coordinates: 482,336,548,322,543,282,478,297,363,316,376,360
        objects:
          - person
      bbq:
        coordinates: 264,349,264,292,101,269,100,349
        objects:
          - person
    motion:
      mask:
        - 149,0,436,0,507,0,564,0,640,0,640,360,590,359,562,231,540,234,500,105,421,43,362,17,329,78,152,11
        - 0,0,0,360,61,360,80,156,129,0
#########################################################################
  Ssol:
    ffmpeg:
      output_args:
        record: preset-record-generic-audio-copy
      inputs:
        - path: rtsp://127.0.0.1:8554/Ssol
          input_args: preset-rtsp-restream
          roles:
            - record
        - path: rtsp://127.0.0.1:8554/SsolSub
          input_args: preset-rtsp-restream
          roles:
            - detect
    detect:
      fps: 6
      width: 480
      height: 360
    objects:
      track:
        - cat
      filters:
        cat:
          min_score: 0.35
          threshold: 0.5
    record:
      enabled: True
      retain:
        days: 7
        mode: motion
      events:
        retain:
          default: 60
          mode: active_objects
        required_zones:
          - litter
    snapshots:
      enabled: true
      timestamp: false
      bounding_box: true
      retain:
        default: 60
      required_zones:
          - litter
    zones:
      litter:
        coordinates: 192,78,192,360,447,360,393,96
        #old 640 = 250,75,244,360,600,360,498,88
        objects:
          - cat
    motion:
      mask:
        - 420,23,420,46,189,46,189,23
#########################################################################
  Ssolfenetre:
    ffmpeg:
      output_args:
        record: preset-record-generic-audio-copy
      inputs:
        - path: rtsp://127.0.0.1:8554/Ssolfenetre
          input_args: preset-rtsp-restream
          roles:
            - record
        - path: rtsp://127.0.0.1:8554/SsolfenetreSub
          input_args: preset-rtsp-restream
          roles:
            - detect
    detect:
      fps: 6
      width: 480
      height: 360
    objects:
      track:
        - person
        - cat
      filters:
        person:
          min_score: 0.35
          threshold: 0.5
        cat:
          min_score: 0.35
          threshold: 0.5
    record:
      enabled: True
      retain:
        days: 7
        mode: motion
      events:
        retain:
          default: 60
          mode: active_objects
        required_zones:
          - mia
          - fenetre
    snapshots:
      enabled: true
      timestamp: false
      bounding_box: true
      retain:
        default: 60
      required_zones:
          - mia
          - fenetre
    zones:
      fenetre:
        coordinates: 346,14,328,183,446,211,464,52
        objects:
          - person
      mia:
        coordinates: 258,45,263,2,174,2,64,19,63,51
        objects:
          - cat
    motion:
      mask:
        - 20,313,255,313,255,336,20,336
################  MQTT #######################

mqtt:
  host: 192.168.1.12
  port: 1883
  topic_prefix: frigate
  client_id: frigate
  user: caublet
  password: BLANKED
  stats_interval: 60

#database:
  # The path to store the SQLite DB (default: shown below)
  #path: /media/frigate/frigate.db
#  path: /db/frigate.db

# Include all cameras by default in Birdseye view
birdseye:
  enabled: True
  mode: continuous
  quality: 8

detectors:
  cpu1:
    type: cpu
    num_threads: 3
  cpu2:
    type: cpu
    num_threads: 3
  cpu3:
    type: cpu
    num_threads: 3

ffmpeg:
  hwaccel_args: preset-nvidia-h264
#  output_args: 
#    record: preset-record-generic-audio-aac
#    record: -f segment -segment_time 10 -segment_format mp4 -reset_timestamps 1 -strftime 1 -c:v copy -c:a aac
#    record: -map 0:v -f segment -segment_time 10 -segment_format mp4 -strftime 1 -c:v h264_nvenc -b:v 4M
#    rtmp:   -vsync 0 -thread_queue_size 256 -rtsp_transport tcp -i rtsp://user:password@ipaddr:port/h264Preview_01_main -map 0 -c:v h264_nvenc -vf format=yuv420p -f segment -segment_time 10 -segment_format mp4 -reset_timestamps 1 -strftime 1 -an
#    detect: -f rawvideo -pix_fmt yuv420p

# Optional: logger verbosity settings
logger:
  # Optional: Default log verbosity (default: shown below)
#  default: info
  default: info
  # Optional: Component specific logger overrides
  logs:
    frigate.event: info
```

### Relevant log output

```shell
none
```

### FFprobe output from your camera

```shell
not needed
```

### Frigate stats

```json
{"Back":{"camera_fps":6.1,"capture_pid":307,"detection_enabled":1,"detection_fps":0.0,"ffmpeg_pid":321,"pid":301,"process_fps":6.1,"skipped_fps":0.0},"Side":{"camera_fps":6.1,"capture_pid":306,"detection_enabled":1,"detection_fps":0.0,"ffmpeg_pid":313,"pid":299,"process_fps":6.1,"skipped_fps":0.0},"Ssol":{"camera_fps":6.1,"capture_pid":311,"detection_enabled":1,"detection_fps":0.0,"ffmpeg_pid":326,"pid":303,"process_fps":6.1,"skipped_fps":0.0},"Ssolfenetre":{"camera_fps":6.0,"capture_pid":316,"detection_enabled":1,"detection_fps":0.0,"ffmpeg_pid":339,"pid":304,"process_fps":6.0,"skipped_fps":0.0},"cpu_usages":{"%Cpu(s):":{"cpu":"id,","mem":"0.0"},"1":{"cpu":"0.0","mem":"0.0"},"108":{"cpu":"0.0","mem":"0.1"},"110":{"cpu":"1.7","mem":"3.8"},"118":{"cpu":"0.0","mem":"0.3"},"122":{"cpu":"0.0","mem":"0.1"},"123":{"cpu":"0.0","mem":"0.1"},"124":{"cpu":"0.0","mem":"0.1"},"125":{"cpu":"0.0","mem":"0.1"},"15":{"cpu":"0.0","mem":"0.0"},"16":{"cpu":"0.0","mem":"0.0"},"24":{"cpu":"0.0","mem":"0.0"},"25":{"cpu":"0.0","mem":"0.0"},"26":{"cpu":"0.0","mem":"0.0"},"27":{"cpu":"0.0","mem":"0.0"},"28":{"cpu":"0.0","mem":"0.0"},"283":{"cpu":"0.0","mem":"2.0"},"289":{"cpu":"0.3","mem":"0.3"},"29":{"cpu":"0.0","mem":"0.0"},"290":{"cpu":"0.0","mem":"2.5"},"292":{"cpu":"0.0","mem":"2.5"},"293":{"cpu":"0.0","mem":"2.5"},"296":{"cpu":"1.0","mem":"2.3"},"299":{"cpu":"0.7","mem":"2.4"},"30":{"cpu":"0.0","mem":"0.0"},"300":{"cpu":"0.0","mem":"0.4"},"301":{"cpu":"0.7","mem":"2.4"},"302":{"cpu":"0.0","mem":"0.4"},"303":{"cpu":"1.0","mem":"2.4"},"304":{"cpu":"0.3","mem":"2.3"},"305":{"cpu":"0.0","mem":"0.4"},"306":{"cpu":"0.3","mem":"2.1"},"307":{"cpu":"0.7","mem":"2.1"},"308":{"cpu":"0.0","mem":"0.4"},"31":{"cpu":"0.0","mem":"0.0"},"311":{"cpu":"0.7","mem":"2.1"},"313":{"cpu":"1.3","mem":"4.2"},"314":{"cpu":"0.0","mem":"0.4"},"316":{"cpu":"0.7","mem":"2.1"},"32":{"cpu":"0.0","mem":"0.0"},"321":{"cpu":"1.3","mem":"4.1"},"325":{"cpu":"1.0","mem":"0.7"},"326":{"cpu":"1.3","mem":"4.1"},"330":{"cpu":"1.0","mem":"0.7"},"338":{"cpu":"0.7","mem":"0.5"},"339":{"cpu":"1.3","mem":"4.1"},"340":{"cpu":"0.3","mem":"0.6"},"343":{"cpu":"0.7","mem":"0.5"},"344":{"cpu":"0.3","mem":"0.6"},"345":{"cpu":"1.0","mem":"0.5"},"346":{"cpu":"1.3","mem":"0.5"},"347":{"cpu":"1.0","mem":"0.5"},"356":{"cpu":"1.3","mem":"0.5"},"370":{"cpu":"1.3","mem":"0.5"},"371":{"cpu":"1.3","mem":"0.5"},"40":{"cpu":"0.0","mem":"0.0"},"43":{"cpu":"0.0","mem":"0.0"},"727":{"cpu":"0.0","mem":"0.0"},"760":{"cpu":"0.3","mem":"0.1"},"762":{"cpu":"0.0","mem":"0.0"},"77":{"cpu":"0.0","mem":"0.0"},"82":{"cpu":"0.0","mem":"0.0"},"83":{"cpu":"0.0","mem":"0.0"},"95":{"cpu":"9.0","mem":"0.5"},"MiB":{"cpu":"1718.4","mem":"avail"},"PID":{"cpu":"%CPU","mem":"%MEM"},"Tasks:":{"cpu":"stopped,","mem":"0"},"top":{"cpu":"users,","mem":"load"}},"detection_fps":0.0,"detectors":{"cpu1":{"detection_start":0.0,"inference_speed":115.79,"pid":290},"cpu2":{"detection_start":0.0,"inference_speed":124.01,"pid":292},"cpu3":{"detection_start":0.0,"inference_speed":124.84,"pid":293}},"gpu_usages":{"NVIDIA GeForce GTX 1050":{"gpu":"0 %","mem":"11.9 %"}},"service":{"last_updated":1682013829,"latest_version":"0.12.0","storage":{"/dev/shm":{"free":63.9,"mount_type":"tmpfs","total":67.1,"used":3.2},"/media/frigate/clips":{"free":162826.3,"mount_type":"ext4","total":324911.6,"used":147745.2},"/media/frigate/recordings":{"free":162826.3,"mount_type":"ext4","total":324911.6,"used":147745.2},"/tmp/cache":{"free":975.8,"mount_type":"tmpfs","total":1000.0,"used":24.2}},"temperatures":{},"uptime":490,"version":"0.12.0-da3e197"}}
```

### Operating system

Proxmox

### Install method

Docker Compose

### Coral version

CPU (no coral)

### Network connection

Wired

### Camera make and model

Anpviz IPC-D352VU-S

**Comment by NickM-27** [comment]
That's incorrect, frigate does not decode recordings they are copied directly from the stream so no encoding is done

**Comment by caublet** [comment]
I can't to find the info on neither : 

- [https://docs.frigate.video/configuration/ffmpeg_presets/#hwaccel-presets](https://docs.frigate.video/configuration/ffmpeg_presets/#hwaccel-presets)
- [https://docs.frigate.video/configuration/hardware_acceleration#nvidia-gpus](https://docs.frigate.video/configuration/hardware_acceleration#nvidia-gpus)

**Comment by caublet** [comment]
> That's incorrect, frigate does not decode recordings they are copied directly from the stream so no encoding is done

Could it be possible to come up with something that could work using : 
`
ffmpeg:
  output_args: 
    record: {That something that need to be worked out}
`

**Comment by NickM-27** [comment]
It's not clear what you're referring to, there is nothing that the GPU can help with. Literally all the record process does is take the camera stream and writes it to RAM, the GPU can't help with that.

**Comment by caublet** [comment]
I've looked on [Video Encode and Decode GPU Support Matrix](https://developer.nvidia.com/video-encode-and-decode-gpu-support-matrix-new) and saw this card had a x264 chip that does encoding. Would like to record using that chip on gpu

**Comment by NickM-27** [comment]
Encoding is only necessary if the stream is decoded, which it is not in frigates case. 

You said 

> The issues comes from when a recording starts; encoding of recording is using CPU

Which is entirely false, there is no encoding being done on cpu or GPU.

**Comment by caublet** [comment]
> Encoding is only necessary if the stream is decoded, which it is not in frigates case.
> 
> You said
> 
> > The issues comes from when a recording starts; encoding of recording is using CPU
> 
> Which is entirely false, there is no encoding being done on cpu or GPU.

Let me rephrase this : "The issues comes from when a recording starts; I see a CPU spike from idle at 30% usage to 300-400% while it's recording"

**Comment by NickM-27** [comment]
That's not going to be from recording. Frigate is always recording from when it starts until it is stopped, it just discards the recording segments that aren't saved. So whatever cpu usage you are seeing is unrelated to recording. It is most likely object detection seeing as you aren't using a coral or other detector accelerator
